src/irmin/tree.ml

```diff
@@ -924,6 +924,18 @@ module Make (P : Backend.S) = struct
       | Error _ as e -> e
       | Ok m -> Ok (StepMap.bindings m)
 
+    let seq_of_updates update_map value_bindings =
+      (* This operation can be costly for a large update_map. *)
+      let value_bindings =
+        Seq.filter (fun (s, _) -> not (StepMap.mem s update_map)) value_bindings
+      in
+      let update_seq =
+        StepMap.to_seq update_map
+        |> Seq.filter_map (fun (s, elt) ->
+               match elt with Remove -> None | Add e -> Some (s, e))
+      in
+      Seq.append value_bindings update_seq
```

Comment:
Perhaps we could go with something like this, which I think is a bit more order-preserving:

```suggestion
    (* Given an unordered sequence of bindings and an update map, return a new
       sequence with the updates applied, preserving the original order of steps. *)
    let rec seq_of_updates updates value_bindings () =
      if StepMap.is_empty updates then
        (* Short-circuit return if we have no more updates to apply *)
        value_bindings ()
      else
        match value_bindings () with
        | Seq.Nil ->
            (* Any remaining [adds] introduce new steps (not part of the
               original order), which we add at the end. *)
            let remainder =
              StepMap.to_seq updates
              |> Seq.filter_map (fun (s, elt) ->
                     match elt with Remove -> None | Add e -> Some (s, e))
            in
            remainder ()
        | Seq.Cons (((step, _) as old_binding), value_bindings) -> (
            match StepMap.find_opt step updates with
            | None ->
                Seq.Cons (old_binding, seq_of_updates updates value_bindings)
            | Some update -> (
                let updates = StepMap.remove step updates in
                let tail = seq_of_updates updates value_bindings in
                match update with
                | Add v' -> Seq.Cons ((step, v'), tail)
                | Remove -> tail ()))
```

Reply:
I'm not sure these extra `StepMap.remove` calls are worth the gain on ordering, especially now that the documentation states that the order is unstable.

Reply:
> I'm not sure these extra `StepMap.remove` calls are worth the gain on ordering

Perhaps not. OTOH, they make the update map smaller (and potentially empty it), so subsequent seq nodes are faster to compute.

> especially now that the documentation states that the order is unstable.

I'm still uncertain about this. There are already at least a few cases where the Tezos protocol depends on the `fold` order exposed by the context (even though it shouldn't), e.g. see here: https://gitlab.com/tezos/tezos/-/issues/944. I expect there to be even more occurrences that expect a stable (though undefined) order. I would be surprised if it's possible to bootstrap a store in which the `fold` order has been randomised, but we should probably try it (and perhaps contribute it as a Tezos test, so we no longer have to worry about changing the order).

Reply:
(Of course, in principle I'd rather either commit to a backend-defined ordering or not make any efforts to maintain one; I'm just worried about breaking existing Tezos code.)

Reply:
Hmm, I spot that @samoht has changed the `fold` in `context.mli` to guarantee a lexical order of `fold` traversal... (https://gitlab.com/tezos/tezos/-/merge_requests/2527/diffs#2f6425c3e68fdbb0fe8a78df1b247b2012a36744_107_107)

Reply:
I think this is no longer the case https://gitlab.com/tezos/tezos/-/blob/master/src/lib_context/context.ml#L293 and https://gitlab.com/tezos/tezos/-/blob/master/src/lib_context/helpers/context.ml#L64, but we are indeed guaranteeing the order over which we fold https://gitlab.com/tezos/tezos/-/blob/master/src/lib_context/sigs/context.ml#L94. And we have indeed tests for folding, I completely forgot about them! (running them now)

Reply:
>  bootstrap a store in which the fold order has been randomised
I've pushed that in my todo :)

We can use your patch in the meantime